Documentation for Flyway support for multiple datasources #3449 (#6086)
quarkusio/quarkus: docs/src/main/asciidoc/flyway.adoc

<1> Inject the Flyway object if you want to use it directly

== Multiple datasources

Flyway can be configured for multiple datasources. 

> It works exactly the same way as for named datasource, for example:

Reviewer: ```suggestion
It works exactly the same way as for the default datasource, for example:
```
?

Author: i reworded the sentence.

quarkus.datasource.users.username=username1
quarkus.datasource.users.min-size=1
quarkus.datasource.users.max-size=11

quarkus.datasource.inventory.driver=org.h2.Driver
quarkus.datasource.inventory.url=jdbc:h2:tcp://localhost/mem:inventory
quarkus.datasource.inventory.username=username2
quarkus.datasource.inventory.min-size=2
quarkus.datasource.inventory.max-size=12

# Flyway configuration for the default datasource
quarkus.flyway.schemas=DEFAULT_TEST_SCHEMA
quarkus.flyway.locations=db/default/location1,db/default/location2
quarkus.flyway.migrate-at-start=true

# Flyway configuration for the "users" datasource
quarkus.flyway.users.schemas=USERS_TEST_SCHEMA
quarkus.flyway.users.locations=db/users/location1,db/users/location2
quarkus.flyway.users.migrate-at-start=true

# Flyway configuration for the "inventory" datasource
quarkus.flyway.inventory.schemas=INVENTORY_TEST_SCHEMA
quarkus.flyway.inventory.locations=db/inventory/location1,db/inventory/location2
quarkus.flyway.inventory.migrate-at-start=true
--

Notice there's an extra bit in the key.
The syntax is as follows: `quarkus.flyway.[optional name.][datasource property]`.

> NOTE: Without configuration, Flyway is setup for every datasource using the default settings.

Reviewer: ```suggestion
NOTE: Without configuration, Flyway is set up for every datasource using the default settings.
```
So... I was wondering if it was a wise choice or if we should enable it only if a location is set? I mean there's very little chance that you will want all your datasources to share the same location? Or am I wrong?

Author: I hear you. I think, it makes sense as it is. 

If there is only one datasource (named or default, doesn't matter) it is perfectly fine, that the default location is used. The most important point IMO is, that "quarkus.flyway.migrate-at-start" is false by default. Nothing "bad" happens, if there are multiple datasources with default Flyway configuration. 

Everything else is in the responsibility of the developer. Even if Flyway would only be activated with a configured location, it would still be possible to copy&paste the same location. It might not be the standard use case, but it could also be used to have two databases with the same scheme (e.g. to synchronize a backup system).

Let me know, if i should start a discussion on google.groups to get more opinions.

Reviewer: Let's try it your way and we'll see what feedback we have.

I let you fix the other things? Thanks!

Author: i did and squashed it into a single commit.

    @Inject
    Flyway flyway; <1>

    @Inject
    @Named("flyway_users") <2>
    Flyway flywayForUsers;

    @Inject

>     @FlywayDataSource("inventory") <3>

Reviewer: Let's make it the first proposal as I would prefer to recommend this usage.

Author: ok

--

<1> Inject the Flyway object if you want to use it directly
<2> Inject Flyway for named datasources

> <3> Inject Flyway for named datasources using the quarkus qualifier

Reviewer: ```suggestion
<3> Inject Flyway for named datasources using the Quarkus `FlywayDataSource` qualifier
```

Author: ok
